### DISCUSSION TITLE
Reverse Engineer the CAPT code to generate a PDF (also LIPS vs CAPT)

### DISCUSSION THREAD

**Discussion by RWAP** [mounaiban/captdriver#37] [Ideas]
This is an odd ask (I realise) but we produce an emulator for various printers - converting the captured data into a PDF so that the printer itself can be removed from the equation completely.

I have only just come across the CAPT printer language and wondered if anyone might be interested in writing a routine to convert the captured data stream (in CAPT) to PDF.

Not easy as the raw data from a Windows test page for the LBP-1310 starts:

00000000 CD CA 10 00 00 11 00 01 00 0D 00 00 00 00 00 00
00000010 00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00

That does not seem to match any of the data I can see in the code!

**Comment by mounaiban** [comment]
There's multiple parts to what you're dealing with here: the LBP-1310, LIPS vs CAPT, the idea of a CAPT-to-PDF conversion and the idea of emulating printers...

#### LBP1310
This device uses a completely different format and protocol, LIPS, according to page 6-4 (PDF p182) the [LBP-1310 manual](https://canon.jp/-/media/Project/Canon/CanonJP/Website/support/manual/lasershot/1310/1310-user.pdf?la=ja-JP&hash=C8DC7172D7ABEB8C8289C821BC1CE1DF). The LBP-1310 seems to be capable of on-device rasterisation, based on the specs on page 6-3 (PDF p181): it has a highly-generalised PowerPC 603ei CPU, as opposed to a simpler and cheaper microcontroller and DSP combo.

This suggests that LIPS might be like PCL, which contains page layout information as well as printer control commands.

So the good news and bad news: you are pretty much ready to use this printer for work as it is [supported by GhostScript](https://www.openprinting.org/driver/lips4/) (and therefore CUPS), but support for the LBP-1310 is beyond the scope of Captdriver.

I'm not fluent in Japanese, so I may be a little wrong, but that's my best understanding for now. Here's a link to the other manuals for the LBP-1310 on Canon's Japan website: https://canon.jp/support/manual/lasershot/1310

#### LIPS vs CAPT
The LBP Image Processing System (LIPS) seems to have a larger scope than CAPT; LIPS includes page description language, CAPT doesn't. LIPS is like a Canon proprietary take on PCL, and appears to be [comprehensively documented](https://canon.jp/-/media/Project/Canon/CanonJP/Website/support/manual/lasershot/1310/1310-lips.pdf?la=ja-JP&hash=6A53BE7A2F6E415EF07662D6C0DFC0A8).

Meanwhile, CAPT was designed to increase the efficiency of host-based rasterisation over a network, in anticipation of an industry standard where servers and workstations would prepare the rasters, then pipe the rasters to the devices in a highly-efficient compressed format. It therefore lacked a page description language; CAPT intended to allow users to choose their own raster image processing solution.

The page description features CAPT misses out on were made up for with more advanced device monitoring and automagic configuration.

#### CAPT-to-PDF
Now we get to a proposed CAPT-to-PDF converter... all you would get from such an endeavour is a PDF that has every page as a single large bitmap. That is because CAPT only handles rasters. :pouting_cat: 

Maybe as a consolation the PDF would contain device status dumps as each page is being printed out?

#### Emulating Printers
I still think emulating the printers in pursuit of device preservation is worth pursuing, if you are attempting something like a MESS/MAME driver that simulates device responses. It'll take a lot more time than a data format converter though...

**Comment by ValdikSS** [comment]
>`CD CA 10`

That's CARPS/UFR-II header.
https://github.com/ondrej-zary/carps-cups/blob/67fb6a543dcdf3beb459d2928f56acab69c78fdb/rastertocarps.c#L34-L36

The data could be anything really. UFR-II can contain JBIG, G4, CAPT or almost-raw image data. CARPS has it's own compression.

**Comment by ValdikSS** [comment]
The driver is here: https://canon.jp/support/software/os/select?pr=134&os=20
It consists of (relevant parts):
* open-source opvp library for ghostscript opvp interface which process the postscript source sent to ghostscript (`Sources/cnrdrvcups-lips4-6.30-1.00.tar.xz/cnrdrvcups-lips4-6.30/driver/`)
* closed-source `cnpkmodulelipsr` (`Sources/cnrdrvcups-lips4-6.30-1.00.tar.xz/lib/libs32/`) which controls protocol framing (`CD CA 10`) and stuff

The opvp ghostscript driver is also used in CAPT for selected printers, it does rasterization. I heard there's also Canon's "vectorization" opvp driver (which operates like PCL), and I guess that's it. The function names `lipsNewPath()`, `lipsStrokePath()`, `lipsBezierPath()` hints that it outputs vector.
